src/net.cpp

```diff
@@ -2635,25 +2635,34 @@ void CConnman::StopNodes()
         }
     }
 
-    // Close sockets
-    LOCK(cs_vNodes);
-    for (CNode* pnode : vNodes)
-        pnode->CloseSocketDisconnect();
-    for (ListenSocket& hListenSocket : vhListenSocket)
-        if (hListenSocket.socket != INVALID_SOCKET)
-            if (!CloseSocket(hListenSocket.socket))
-                LogPrintf("CloseSocket(hListenSocket) failed with error %s\n", NetworkErrorString(WSAGetLastError()));
+    // Delete peer connections
+    std::vector<CNode*> nodes;
+    {
+        LOCK(cs_vNodes);
+        nodes = std::move(vNodes);
+        vNodes.clear();
```

Comment:
Could achieve the same in one less line of code with https://en.cppreference.com/w/cpp/container/vector/swap ?

Reply:
Or even remove all of this logic and re-use the existing "disconnect all nodes" logic?

```diff
diff --git a/src/net.cpp b/src/net.cpp
index b7c1b8c6c4..9a4a989e15 100644
--- a/src/net.cpp
+++ b/src/net.cpp
@@ -2636,31 +2636,17 @@ void CConnman::StopNodes()
     }
 
     // Delete peer connections
-    std::vector<CNode*> nodes;
     {
         LOCK(cs_vNodes);
-        nodes = std::move(vNodes);
-        vNodes.clear();
+        assert(vNodes.empty());
     }
-
-    for (CNode* pnode : nodes) {
-        pnode->CloseSocketDisconnect();
-        DeleteNode(pnode);
-    }
-
-    for (CNode* pnode : vNodesDisconnected) {
-        DeleteNode(pnode);
-    }
-    vNodesDisconnected.clear();
+    assert(vNodesDisconnected.empty());
 
     // Close listening sockets
-    for (ListenSocket& hListenSocket : vhListenSocket) {
-        if (hListenSocket.socket != INVALID_SOCKET) {
-            if (!CloseSocket(hListenSocket.socket)) {
+    for (ListenSocket& hListenSocket : vhListenSocket)
+        if (hListenSocket.socket != INVALID_SOCKET)
+            if (!CloseSocket(hListenSocket.socket))
                 LogPrintf("CloseSocket(hListenSocket) failed with error %s\n", NetworkErrorString(WSAGetLastError()));
-            }
-        }
-    }
     vhListenSocket.clear();
 
     semOutbound.reset();
diff --git a/src/net.h b/src/net.h
index 13f86b6a91..5a87a5a304 100644
--- a/src/net.h
+++ b/src/net.h
@@ -867,6 +867,8 @@ public:
     void Stop()
     {
         StopThreads();
+fNetworkActive=true;
+DisconnectNodes();
         StopNodes();
     };

Reply:
(I tried running the tests with my diff and for some reason the node crashed)

Reply:
> Could achieve the same in one less line of code with https://en.cppreference.com/w/cpp/container/vector/swap ?

Thanks! [Updated](https://github.com/bitcoin/bitcoin/pull/21563#issuecomment-824898734).

Reply:
Just to answer (myself) why my diff crashed: When the threads are interrupted, the node refcounts do *not* get decremented

Reply:
I am working on a change that will address this problem of forgetting `Release()` due to an early `return`.

Reply:
> When the threads are interrupted, the node refcounts do _not_ get decremented

This is changed/fixed in https://github.com/bitcoin/bitcoin/pull/21878 in commit

96bcc72b5 `net: dedup and RAII-fy the creation of a copy of CConnman::vNodes`